### ISSUE TITLE
Optimisation assertion error for Pydantic library

### ISSUE THREAD

**Issue by coderfromhere** [Nuitka/Nuitka#635]
Hi, I'm getting the following assertion error around optimizer while compiling a project that uses [Pydantic](https://github.com/samuelcolvin/pydantic)

The System
```
MacOS Catalina
Python 3.7.6 (default, Jan  6 2020, 14:21:53) [Clang 7.1.0 (tags/RELEASE_710/final)] on darwin
```

Nuitka build:
```
0.6.7
Python: 3.7.6 (default, Jan  6 2020, 14:21:53) 
Executable: /Users/project/.venv/bin/python
OS: Darwin
Arch: x86_64
```

The environment with Python is created by the following [Nix shell](https://nixos.org/nix/manual/#idm140737322627776):
```nix
with (import (builtins.fetchGit {
    # https://nixos.wiki/wiki/FAQ/Pinning_Nixpkgs
    name = "nuitka-python37";
    url = https://github.com/nixos/nixpkgs/;
    ref = "refs/heads/master";
    rev = "6f545410b593a0d1b78806aa878f95ee3f3d4a4a";
}) {});

mkShell {
    name = "nuitka-python37-python37";
    buildInputs = [
        # see https://nixos.org/nixos/packages.html
        # Python distribution
        python37Full
        python37Packages.virtualenv
        poetry
        # usually required as python runtime dependencies
        ncurses
        libxml2
        libxslt
        libzip
        zlib
        # root CA certificates
        cacert
        gnumake
        which
    ];
    shellHook = ''
        # Set SOURCE_DATE_EPOCH so that we can use python wheels.
        # This compromises immutability, but is what we need
        # to allow package installs from PyPI
        export SOURCE_DATE_EPOCH=$(date +%s)

        VENV_DIR=$PWD/.venv

        export PATH=$VENV_DIR/bin:$PATH
        export PYTHONPATH=""
        export LANG=en_US.UTF-8

        # https://python-poetry.org/docs/configuration/
        export POETRY_VIRTUALENVS_CREATE=true
        export POETRY_VIRTUALENVS_IN_PROJECT=true
        export PIP_CACHE_DIR=$PWD/.local/pip-cache

        # Setup virtualenv
        if [ ! -d $VENV_DIR ]; then
            virtualenv $VENV_DIR
        fi

        # Dirty fix for Linux systems
        # https://nixos.wiki/wiki/Packaging/Quirks_and_Caveats
        export LD_LIBRARY_PATH=${stdenv.cc.cc.lib}/lib/:$LD_LIBRARY_PATH
    '';
}
```

The traceback is:
```
Nuitka:INFO:Optimizing module 'pydantic', 13 more modules to go after that.
Nuitka:INFO:Doing module local optimizations for 'pydantic'.
Traceback (most recent call last):
  File "/Users/project/.venv/lib/python3.7/site-packages/nuitka/__main__.py", line 182, in <module>
    main()
  File "/Users/project/.venv/lib/python3.7/site-packages/nuitka/__main__.py", line 175, in main
    MainControl.main()
  File "/Users/project/.venv/lib/python3.7/site-packages/nuitka/MainControl.py", line 736, in main
    main_module = createNodeTree(filename=filename)
  File "/Users/project/.venv/lib/python3.7/site-packages/nuitka/MainControl.py", line 142, in createNodeTree
    Optimization.optimize(main_module.getOutputFilename())
  File "/Users/project/.venv/lib/python3.7/site-packages/nuitka/optimizations/Optimization.py", line 619, in optimize
    finished = makeOptimizationPass(initial_pass=False)
  File "/Users/project/.venv/lib/python3.7/site-packages/nuitka/optimizations/Optimization.py", line 530, in makeOptimizationPass
    changed = optimizeModule(current_module)
  File "/Users/project/.venv/lib/python3.7/site-packages/nuitka/optimizations/Optimization.py", line 168, in optimizeModule
    changed = optimizeCompiledPythonModule(module)
  File "/Users/project/.venv/lib/python3.7/site-packages/nuitka/optimizations/Optimization.py", line 127, in optimizeCompiledPythonModule
    Plugins.considerImplicitImports(module=module, signal_change=signalChange)
  File "/Users/project/.venv/lib/python3.7/site-packages/nuitka/plugins/Plugins.py", line 141, in considerImplicitImports
    plugin.considerImplicitImports(module, signal_change)
  File "/Users/project/.venv/lib/python3.7/site-packages/nuitka/plugins/PluginBase.py", line 207, in considerImplicitImports
    signal_change=signal_change,
  File "/Users/project/.venv/lib/python3.7/site-packages/nuitka/plugins/PluginBase.py", line 498, in recurseTo
    reason=reason,
  File "/Users/project/.venv/lib/python3.7/site-packages/nuitka/importing/Recursion.py", line 167, in recurseTo
    reason=reason,
  File "/Users/project/.venv/lib/python3.7/site-packages/nuitka/importing/Recursion.py", line 83, in _recurseTo
    is_shlib=module_kind == "shlib",
  File "/Users/project/.venv/lib/python3.7/site-packages/nuitka/tree/Building.py", line 924, in decideModuleTree
    source_ref=source_ref,
  File "/Users/project/.venv/lib/python3.7/site-packages/nuitka/nodes/ModuleNodes.py", line 264, in __init__
    setLocalsDictType(self.module_dict_name, "module_dict")
  File "/Users/project/.venv/lib/python3.7/site-packages/nuitka/nodes/LocalsScopes.py", line 31, in setLocalsDictType
    assert locals_dict_name not in locals_dict_handles, locals_dict_name
AssertionError: globals_pydantic$__init__
```

The application:
```python
import pydantic


if __name__ == "__main__":
    print('Hello world!')
```

Build script:
```
python -m nuitka --show-progress --follow-imports ./nuitka-app.py
```

**Comment by kayhayen** [comment]
I am getting the similar import error for these ones as in #689 and I think pydantic has worked before, so there must also be a regression in that.

**Comment by kayhayen** [comment]
So, my bisect has lead to me to the point, where I moved the assignment as the Nuitka meta path loader as responsible to sooner. Before it was `FileFinder` until being replaced.

Somebody has decided to compile all of pydantic into extension modules, which is why nothing is compiled except the top level package. The extension modules win over `.py` files in scan, so there is of course a question, wether to make an exception for these ones. I appreciate the pure Python source code being present.

In order to reproduce your problem, I will need to get that addressed, and to be compatible, the new loader is still the best idea, and not really a regression in that sense.

**Comment by kayhayen** [comment]
So, I am just pushing the solution for the non-compilation to factory, and when I re-compile everything it works, and I believe there are no possible changes since last release that may have caused a change for the optimization error you saw.

Can you give me the pydantic version from `python -m pip freeze` that you were using at the time, then I could try that one out.

However, with current factory, it soon ought to work.

**Comment by coderfromhere** [comment]
Hi @kayhayen, thanks for looking at this. The pydantic version I tried was `pydantic==1.4`

**Comment by kayhayen** [comment]
So this one also is all extension modules on me now, too, leading to no compilation at all. With the ```--prefer-source-code`` option we are getting also no error anymore.

And now I recall it was a problem with imports not being understood properly, leading to some form of duplication and this was solved in #618 long time ago, and with 0.6.8 was released.
